File: TP1/RushHour.java
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.PriorityQueue;
import java.util.Queue;

public class RushHour {

	/*
	 * representation du probleme : Les six lignes sont numerotees de haut en
	 * bas, de 0 a 5 et les 6 colonnes de gauche a droite, de 0 a 5.
	 *
	 * il y a nbcars voitures, numérotées de 0 a nbcars-1 pour chaque voiture i
	 * on a : - color[i] sa couleur - horiz[i] si la voiture est horizontale
	 * (vrai) ou verticale (faux) - len[i] sa longueur (2 ou 3) - moveon[i] la
	 * ligne (si horiz[i]) ou la colonne (sinon) où se trouve la voiture
	 *
	 */

	public int nbcars;
	public String[] color;
	public boolean[] horiz;
	public int[] len;
	public int[] moveon;

	public int nbMoves;
	/*
	 * la matrice free permet de savoir si une case est libre
	 */
	public boolean[][] free = new boolean[6][6];

	void initFree(State s) {
		//TODO
	}

	/*
	 * renvoie la liste des déplacements possibles
	 */

	public ArrayList<State> moves(State s) {
		initFree(s);
		ArrayList<State> l = new ArrayList<State>();
		//TODO
		return l;
	}

	/*
	 * trouve une solution à partir de s
	 */
	public State solve(State s) {
		HashSet<State> visited = new HashSet<State>();
		visited.add(s);
		Queue<State> Q = new LinkedList<State>();
		//TODO
		System.out.println("pas de solution");
		return null;
	}

	public State solveAstar(State s) {
		HashSet<State> visited = new HashSet<State>();
		visited.add(s);
		PriorityQueue<State> Q = new PriorityQueue<State>(10, new MyComparator());
		//TODO
		System.out.println("pas de solution");
		return null;
	}

	/*
	 * affiche la solution
	 */

	void printSolution(State s) {
		// TODO
	}
	
	private class MyComparator implements Comparator<State> {
		@Override
		public int compare(State arg0, State arg1) {
			return arg0.f - arg1.f;
		}
	}
}

File: TP1/rushHour.py
from priorityQueue import PriorityQueue
from state import State
import time


class RushHour(object):
    
    def __init__(self):
        """
            definition de la grille
        """
        self.nbcars = 0
        self.color = []
        self.horiz = []
        self.len = []
        self.moveOn = []
        self.free = [[True] * 6 for _ in range(6)]

    def numberOfCarsBlocking(self, state):
        free = [[True] * 6 for _ in range(6)]
        for i in range(self.nbcars):
            if self.horiz[i]:
                x = self.moveOn[i]
                y = state.pos[i]
                for j in range(self.len[i]):
                    free[x][y+j] = False
            else:
                y = self.moveOn[i]
                x = state.pos[i]
                for j in range(self.len[i]):
                    free[x+j][y] = False
        return len([x for x in free[1][state.pos[0]+self.len[0]-1:] if not x])

    def initFree(self, state):
        self.free = [[True] * 6 for _ in range(6)]
        for i in range(self.nbcars):
            if self.horiz[i]:
                x = self.moveOn[i]
                y = state.pos[i]
                for j in range(self.len[i]):
                    self.free[x][y+j] = False
            else:
                y = self.moveOn[i]
                x = state.pos[i]
                for j in range(self.len[i]):
                    self.free[x+j][y] = False

    def moves(self, state):
        self.initFree(state)
        l = []
        for i in range(self.nbcars):
            if self.horiz[i]:
                x = self.moveOn[i]
                y = state.pos[i]
                if y > 0 and self.free[x][y-1]:
                    l.append(State(state, i, -1))
                y += self.len[i] - 1
                if y < 5 and self.free[x][y+1]:
                    l.append(State(state, i, 1))
            else:
                y = self.moveOn[i]
                x = state.pos[i]
                if x > 0 and self.free[x-1][y]:
                    l.append(State(state, i, -1))
                x += self.len[i] - 1
                if x < 5 and self.free[x+1][y]:
                    l.append(State(state, i, 1))
        return l

    def solve(self,state):
        start = time.time()
        visited = set()
        visited.add(state)
        q = self.moves(state)
        while q:
            curr = q.pop(0)
            if curr.success():
                print("Nombre d'états visités : " + str(len(visited)))
                print("Temps écoulé : " + '%.3f' % (time.time() - start) + 's')
                return curr
            nexts = self.moves(curr)
            for s in nexts:
                if not s in visited:
                    q.append(s)
                visited.add(s)
        print("Pas de solution")
        return None
    
    def solveAstar(self,state):
        start = time.time()
        visited = set()
        visited.add(state)
        q = PriorityQueue()
        initialMoves = self.moves(state)
        for s in initialMoves:
            q.enqueue(s, s.f)
        while q:
            curr = q.dequeue()
            if curr.success():
                print("Nombre d'états visités : " + str(len(visited)))
                print("Temps écoulé : " + '%.3f' % (time.time() - start) + 's')
                return curr
            nexts = self.moves(curr)
            for s in nexts:
                if not s in visited:
                    q.enqueue(s, s.f)
                visited.add(s)
        print("Pas de solution")
        return None

    def printSolution(self,state):
        res = []
        if state is None:
            return
        while state.prev is not None:
            aux = "Voiture " + self.color[state.c] + " vers "
            if(self.horiz[state.c]):
                if state.d == 1:
                    aux += "la droite"
                else:
                    aux += "la gauche"
            else:
                if state.d == 1:
                    aux += "le bas"
                else:
                    aux += "le haut"
            res.append(aux)
            state = state.prev
        print(str(len(res)) + " mouvements")
        for string in res[::-1]:
            print(string)

    def printAnimatedAsciiSolution(self, state):
        import sys 
        res = []
        if state is None:
            return
        while state.prev is not None:
            res.append(state)
            state = state.prev
        for s in res[:0:-1]:
            print(s)
            for i in range(7):
                sys.stdout.write("\033[F") # Cursor up one line
            time.sleep(0.33)
        print(res[0])